lib/sources.nix

```diff
@@ -101,4 +101,201 @@ rec {
+  #
+  # Example: For the tree
+  #
+  #     a
+  #       b-leaf
+  #       c-leaf
+  #       d
+  #         e-leaf
+  #     f-leaf
+  #
+  # represented as attrset
+  #
+  #     { a = { b = null; c = null; d = { e = null; }; }; f = null; }
+  #
+  # we have
+  #
+  #     isPrefixOfLeafPath [a] == true
+  #     isPrefixOfLeafPath [x] == false
+  #     isPrefixOfLeafPath [a b] == true
+  #     isPrefixOfLeafPath [a b c] == true
+  #     isPrefixOfLeafPath [a b c x] == true
+  #     isPrefixOfLeafPath [a d] == false
+  isPrefixOfLeafPath = path: tree:
+    if tree == null
+      then true
+      else
+        if path == []
+          then false
+          else
+            let
+              component = builtins.head path;
+              restPath = builtins.tail path;
+            in
+              if !(builtins.hasAttr component tree)
+                then false
+                else
+                  let
+                    subtree = builtins.getAttr component tree;
+                  in
+                    isPrefixOfLeafPath restPath subtree;
+
+
+
+  # See `explicitFilterSource` for an example of this this filter.
+  #
+  # You can use this filter standalone (with `builtins.filterSource`
+  # or better, `builtins.path` with explicitly given `name`) when
+  # you want to combine it with other filters.
+  explicitSourceFilter =
+    {
+      # List of dirs under which all recursively contained files are taken in
+      # (unless a file is filtered by other arguments).
+      # Dirs that match explicitly are immediately taken in.
+      srcDirs ? [],
+      # Explicit list of files that should be taken in.
+      srcFiles ? [],
+      # Exclude dotfiles/dirs by default (unless they are matched explicitly)?
+      excludeHidden ? true,
+      # If any of the path components given here appears anywhere in the path,
+      # (e.g. X in `.../X/...`), the path is excluded (unless matched explicitly).
+      # Example: `pathComponentExcludes = ["gen", "build"]`.
+      pathComponentExcludes ? [],
+
+      # Debugging
+
+      # Enable this to enable a `builtins.trace` output that prints which files
+      # were matched as source inputs.
+      # Output looks like:
+      #     trace: myproject: include regular   /home/user/myproject/Setup.hs
+      #     trace: myproject: skip    regular   /home/user/myproject/myproject.nix
+      #     trace: myproject: skip    directory /home/user/myproject/dist
+      #     trace: myproject: include directory /home/user/myproject/images
+      #     trace: myproject: include regular   /home/user/myproject/images/image.svg
+      #     trace: myproject: include directory /home/user/myproject/src
+      #     trace: myproject: include directory /home/user/myproject/src/MyDir
+      #     trace: myproject: include regular   /home/user/myproject/src/MyDir/File1.hs
+      #     trace: myproject: include regular   /home/user/myproject/src/MyDir/File2.hs
+      debugEnableTracing ? false,
```

Comment:
Maybe `debugTraceEnable`, aligning it with `debugTracePrefix`

Reply:
Done.